central/group/datastore/internal/store/store_impl.go

```diff
@@ -61,16 +63,23 @@ func (s *storeImpl) Walk(authProviderID string, attributes map[string][]string)
 
 	// Search for groups in the list.
 	err = s.db.View(func(tx *bolt.Tx) error {
-		buc := tx.Bucket(groupsBucket)
 		for _, check := range toSearch {
-			serializedKey := serializePkg.PropsKey(check)
-			if serializedVal := buc.Get(serializedKey); serializedVal != nil {
-				grp, err := deserialize(serializedKey, serializedVal)
-				if err != nil {
-					return err
+			grpss, err := s.GetFiltered(func(props *storage.GroupProperties) bool {
```

Comment:
This changes the time complexity of `Walk()` and hence of `storeBasedMapperImpl.getRoles()` from `O(k * log n)` to `O(k * n)`. It's hard to believe that this would be a drastic change but it feels that the whole composite key jazz with serializing and deserializing was done for that logarithmic search. Maybe it's worth verifying with @viswajithiii?

I think once we migrate to postgres, we might be able to create arbitrary indices and so fix this back to logarithm. Could you please check with @connorgorman and @rhybrillou?

Reply:
This is totally fine, I don't think we expect there to be enough groups for this to be a big deal. And you're right that postgres will make this kind of moot. Fine with me to go ahead with this. (Note: commenting based on the description of the change, haven't looked at the code itself)

Reply:
@rukletsov I've talked to Yann, doing things like that is planned based on the performance investigation results. Right now this would be later this year, where this place would be an opportunity for improvement (albeit, as discussed, the impact may not be as big as other places). Potentially @connorgorman  will have more info, but I don't think it will block us in this PR.

Reply:
If reading groups is a hot path, we can just cache all of them in memory. This seems like a store with a small cardinality so walk is probably fine for now